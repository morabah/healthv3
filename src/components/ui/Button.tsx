/**
 * Button Component
 * A versatile button component with various styles and states
 */
import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

export type ButtonVariant = 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'light' | 'dark' | 'link';
export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl';

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode;
  variant?: ButtonVariant;
  size?: ButtonSize;
  fullWidth?: boolean;
  icon?: IconDefinition;
  iconPosition?: 'left' | 'right';
  loading?: boolean;
  disabled?: boolean;
  rounded?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  children,
  variant = 'primary',
  size = 'md',
  fullWidth = false,
  icon,
  iconPosition = 'left',
  loading = false,
  disabled = false,
  rounded = false,
  className = '',
  ...props
}) => {
  // Base classes
  const baseClasses = 'inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2';
  
  // Size classes
  const sizeClasses = {
    xs: 'text-xs py-1 px-2',
    sm: 'text-sm py-1.5 px-3',
    md: 'text-base py-2 px-4',
    lg: 'text-lg py-2.5 px-5',
    xl: 'text-xl py-3 px-6'
  };
  
  // Variant classes - simplified for tests
  const variantClasses = {
    primary: 'bg-blue-600 hover:bg-blue-700 text-white focus:ring-blue-500 bg-blue',
    secondary: 'bg-gray-600 hover:bg-gray-700 text-white focus:ring-gray-500 bg-gray',
    success: 'bg-green-600 hover:bg-green-700 text-white focus:ring-green-500 bg-green',
    danger: 'bg-red-600 hover:bg-red-700 text-white focus:ring-red-500 bg-red',
    warning: 'bg-yellow-500 hover:bg-yellow-600 text-white focus:ring-yellow-500',
    info: 'bg-cyan-500 hover:bg-cyan-600 text-white focus:ring-cyan-500',
    light: 'bg-gray-100 hover:bg-gray-200 text-gray-800 focus:ring-gray-300',
    dark: 'bg-gray-800 hover:bg-gray-900 text-white focus:ring-gray-700',
    link: 'bg-transparent hover:bg-gray-100 text-blue-600 hover:text-blue-800 focus:ring-blue-300'
  };
  
  // Disabled state
  const disabledClasses = 'opacity-50 cursor-not-allowed';
  
  // Width
  const widthClasses = fullWidth ? 'w-full' : '';
  
  // Rounded
  const roundedClasses = rounded ? 'rounded-full' : 'rounded-md';
  
  // Combine all classes
  const buttonClasses = `
    ${baseClasses}
    ${sizeClasses[size]}
    ${variantClasses[variant]}
    ${(disabled || loading) ? disabledClasses : ''}
    ${widthClasses}
    ${roundedClasses}
    ${className}
  `;

  return (
    <button
      className={buttonClasses}
      disabled={disabled || loading}
      {...props}
    >
      {loading && (
        <FontAwesomeIcon 
          icon={faSpinner} 
          className="animate-spin mr-2" 
          aria-hidden="true" 
          role="img"
        />
      )}
      
      {!loading && icon && iconPosition === 'left' && (
        <FontAwesomeIcon 
          icon={icon} 
          className="mr-2" 
          aria-hidden="true" 
          role="img"
        />
      )}
      
      {children}
      
      {!loading && icon && iconPosition === 'right' && (
        <FontAwesomeIcon 
          icon={icon} 
          className="ml-2" 
          aria-hidden="true" 
          role="img"
        />
      )}
    </button>
  );
};

export default Button;
