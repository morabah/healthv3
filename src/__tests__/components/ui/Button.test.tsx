import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import Button from '@/components/ui/Button';
import { faUser } from '@fortawesome/free-solid-svg-icons';

describe('Button Component', () => {
  it('renders correctly with default props', () => {
    const { getByRole } = render(<Button>Click me</Button>);
    const button = getByRole('button', { name: /click me/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass('bg-blue');
  });

  it('renders with different variants', () => {
    const { getByRole, rerender } = render(<Button variant="secondary">Secondary</Button>);
    expect(getByRole('button')).toHaveClass('bg-gray');

    rerender(<Button variant="success">Success</Button>);
    expect(getByRole('button')).toHaveClass('bg-green');

    rerender(<Button variant="danger">Danger</Button>);
    expect(getByRole('button')).toHaveClass('bg-red');
  });

  it('renders with different sizes', () => {
    const { getByRole, rerender } = render(<Button size="sm">Small</Button>);
    expect(getByRole('button')).toHaveClass('text-sm');

    rerender(<Button size="lg">Large</Button>);
    expect(getByRole('button')).toHaveClass('text-lg');
  });

  it('renders with an icon', () => {
    const { getByRole } = render(<Button icon={faUser}>With Icon</Button>);
    const icon = getByRole('img', { hidden: true });
    expect(icon).toBeInTheDocument();
  });

  it('renders in loading state', () => {
    const { getByRole } = render(<Button loading>Loading</Button>);
    const spinner = getByRole('img', { hidden: true });
    expect(spinner).toBeInTheDocument();
  });

  it('calls onClick handler when clicked', async () => {
    const handleClick = jest.fn();
    const { getByRole } = render(<Button onClick={handleClick}>Click me</Button>);
    const button = getByRole('button');
    await userEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('does not call onClick when disabled', async () => {
    const handleClick = jest.fn();
    const { getByRole } = render(<Button disabled onClick={handleClick}>Disabled</Button>);
    const button = getByRole('button');
    await userEvent.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('renders as full width when fullWidth prop is true', () => {
    const { getByRole } = render(<Button fullWidth>Full Width</Button>);
    expect(getByRole('button')).toHaveClass('w-full');
  });

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(<Button>Snapshot Test</Button>);
    expect(container).toMatchSnapshot();
  });
});
