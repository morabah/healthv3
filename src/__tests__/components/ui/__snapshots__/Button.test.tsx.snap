// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<div>
  <button
    class="
    inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2
    text-base py-2 px-4
    bg-blue-600 hover:bg-blue-700 text-white focus:ring-blue-500 bg-blue
    
    
    rounded-md
    
  "
  >
    Snapshot Test
  </button>
</div>
`;
